File: insertion_sort/insertion_sort.go
package insertion_sort

func InsertionSort(arr []int) []int {
    if len(arr) <= 1 {
        return arr
    }
	sorted := make([]int, 0, len(arr))
	sorted = append(sorted, arr[0])
	var res []int
	for i := 1; i < len(arr); i++ {
		res = doInsert(arr[i], sorted)
	}

	return res
}

func doInsert(num int, sorted []int) []int {
	newSorted := make([]int, 0, len(sorted)+1)
	for i := 0; i < len(sorted); i++ {
		currentValue := sorted[i]
		if num >= currentValue {
			newSorted = append(newSorted, currentValue)
		} else {
			newSorted = append(newSorted, []int{num, currentValue}...)
		}
	}

	return newSorted
}

File: insertion_sort/insertion_sort.exs
defmodule InsertionSort do
  @doc """
    Receive a list of integer, sort it using insertion sort algorithm,
    return {:ok, result} | {:error, :not_a_list} | {:error, :not_integer}
  """
  @spec sort(list(integer())) ::
          {:ok, list(integer())} | {:error, :not_a_list} | {:error, :not_integer}
  def sort(arr) when not is_list(arr), do: {:error, :not_a_list}
  def sort([head | _]) when not is_integer(head), do: {:error, :not_integer}

  def sort(arr) do
    do_sort(arr, [])
  end

  defp do_sort([], sorted), do: {:ok, sorted}

  defp do_sort([head | tail], sorted) do
    do_sort(tail, insert(head, sorted))
  end

  defp insert(head, []), do: [head]

  defp insert(x, [y | ys]) do
    if x < y do
      [x, y | ys]
    else
      [y | insert(x, ys)]
    end
  end
end

ExUnit.start()

defmodule TestInsertionSort do
  use ExUnit.Case, async: true

  test "insertion sort 1" do
    assert InsertionSort.sort([7, 12, 9, 11, 3]) === {:ok, [3, 7, 9, 11, 12]}
  end

  test "insertion sort 2" do
    assert InsertionSort.sort([2, 1, 1, 3]) === {:ok, [1, 1, 2, 3]}
  end

  test "empty list" do
    assert InsertionSort.sort([]) === {:ok, []}
  end

  test "length one" do
    assert InsertionSort.sort([1]) === {:ok, [1]}
  end

  test "not a list" do
    assert InsertionSort.sort(123) === {:error, :not_a_list}
  end

  test "not integer" do
    assert InsertionSort.sort(["a", "c"]) === {:error, :not_integer}
  end
end
